### PR TITLE
Add Custom Operation Test in azure/core/basic

### PR DESCRIPTION
**Pull request by mcgallan**
The purpose of this PR is to migrate [the tests](https://github.com/Azure/autorest.csharp/blob/aee28deecf34b25e50cfdc950d572be6fb9aa6a8/test/TestProjects/Authoring-TypeSpec/Authoring-TypeSpec.tsp#L290) from the test.project folder in autorest.csharp to the cadl branch, specifically referring to the tests in latestVersion.typeSpec as detailed in the .changeset.
related issue: https://github.com/Azure/autorest.csharp/issues/4848

### CODE REVIEW

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 1]
```diff
@@ -70,6 +70,20 @@ model UserExportParams {
...
 
+model SwapUsersOptions {
+  @doc("Represents the first user id.")
+  firstUserId: int32;
+
+  @doc("Represents the second user id.")
+  secondUserId: int32;
+}
+
+model SwapUsersParams {
```
Comment (pshao25) [inline packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:81]: How is this test related to "swap user"?

Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: I have made improvements to this method, instead of simply calling it for the user. Regarding the swapParameter, I think it serves as a credential indicating which two models to swap, and it cannot be modified for the time being.

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 2]
```diff
@@ -214,3 +238,16 @@ Expected response body:
...
+@scenario
+@scenarioDoc("""
+Should only generate models named User and UserOrder.
+Expected query parameter: api-version=2022-12-01-preview
+""")
+@collectionAction(Dog, "swapdogs")
+op swapdogs is ResourceOperations.ResourceCollectionAction<
+  Dog,
+  SwapDogsParams,
+  global.Azure.Core.Foundations.AcceptedResponse
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:252]: Why use Foundations? There are not recommended to be used by Azure lib.

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 3]
```diff
@@ -214,3 +238,16 @@ Expected response body:
...
 op export is ResourceOperations.ResourceAction<User, UserExportParams, User>;
+
+/** Swaps two existing dogs with each other. */
+@scenario
+@scenarioDoc("""
+Should only generate models named User and UserOrder.
+Expected query parameter: api-version=2022-12-01-preview
+""")
+@collectionAction(Dog, "swapdogs")
+op swapdogs is ResourceOperations.ResourceCollectionAction<
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:249]: swapdogs...

Could we use test with more meaningful API?

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 2 above)
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: fixed

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 3 above)
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: fixed

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 4]
```diff
@@ -214,3 +216,41 @@ Expected response body:
...
+  },
+  {
+    "id": 2,
+    "name": "John",
+    "etag": "22bdc430-65e8-45ad-81d9-8ffa60d55b59"
+  }
+]
+""")
+@collectionAction(User, "exportallusers")
+op exportAllUsers is ResourceOperations.ResourceCollectionAction<
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:249]: Why do you need `@collectionAction(User, "exportallusers")`? I think it should be included in `ResourceCollectionAction` operation template.

See if you can add a `@post` above the operation. This would be redundant, but the summary seems messed-up without it https://github.com/Azure/cadl-ranch/pull/625/files#diff-e96e1cba09706c8a0e00bde73f721986543b99e3a0db21b3bbb98559db31185a

This is not required. But the summary seems to be

**Review by weidongxu-microsoft — COMMENTED** [review]
@pshao25 for double check.

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 5]
```diff
@@ -214,3 +216,41 @@ Expected response body:
...
+    "id": 1,
+    "name": "Madge",
+    "etag": "11bdc430-65e8-45ad-81d9-8ffa60d55b59"
+  },
+  {
+    "id": 2,
+    "name": "John",
+    "etag": "22bdc430-65e8-45ad-81d9-8ffa60d55b59"
+  }
+]
```
[the remark this answers is not in the document]
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: fixed

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 4 above)
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: Based on the generated results, it is necessary to retain @collectionAction. After removing it, the generated path directly calls the operation name, resulting in uppercase letters appearing in the path. Therefore, I added @collectionAction to prevent this from happening. @post added.

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 5 above)
[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: It's not fixed. Now is it not even a valid json. Please do a simple verify.

[the remark this answers is not in the document]
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: The corrections have been made, including fixing the case issue with "users".

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 4 above)
Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: Eh, who says uppercase letters cannot appear in the path?

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp [hunk 6]
```diff
@@ -214,3 +216,44 @@ Expected response body:
...
+      "name": "Madge",
+      "etag": "11bdc430-65e8-45ad-81d9-8ffa60d55b59"
+    },
+    {
+      "id": 2,
+      "name": "John",
+      "etag": "22bdc430-65e8-45ad-81d9-8ffa60d55b59"
+    }
+  ]
+}
```
Comment (weidongxu-microsoft) [inline packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:248]: you miss the ending ```

Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp:248]: fixed

File: packages/cadl-ranch-specs/http/azure/core/basic/main.tsp (hunk 4 above)
Comment (mcgallan) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: In a previous PR, there was a comment suggesting that I change the uppercase letters in the path to lowercase. Additionally, I have rarely seen paths with uppercase letters in mockapi. So, which approach is more appropriate?

Comment (weidongxu-microsoft) [reply packages/cadl-ranch-specs/http/azure/core/basic/main.tsp]: https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#http-url-casing

I don't have particular preference in this case. It is fine for all-lower-case. Just want to know why the extra `@collectionAction`.

**Review by weidongxu-microsoft — APPROVED** [review]
Please wait another approval (from non-Java dev).

**Review by ArcturusZhang — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
@mcgallan 

Please format the PR again.

**Comment by weidongxu-microsoft** [comment]
Format still fails
